MoltenVK/MoltenVK/GPUObjects/MVKImage.mm

```diff
@@ -716,6 +711,40 @@ static MTLRegion getMTLRegion(const ImgRgn& imgRgn) {
 	return VK_SUCCESS;
 }
 
+VkResult MVKImage::copyImageToMemory(const VkCopyImageToMemoryInfoEXT* pCopyImageToMemoryInfo) {
+#if MVK_MACOS
+	// On macOS, if the device doesn't have unified memory, and the texture is using managed memory, we need
+	// to sync the managed memory from the GPU, so the texture content is accessible to be copied by the CPU.
+	if ( !getPhysicalDevice()->getHasUnifiedMemory() && getMTLStorageMode() == MTLStorageModeManaged ) {
+		@autoreleasepool {
+			id<MTLCommandBuffer> mtlCmdBuff = getDevice()->getAnyQueue()->getMTLCommandBuffer(kMVKCommandUseCopyImageToMemory);
+			id<MTLBlitCommandEncoder> mtlBlitEnc = [mtlCmdBuff blitCommandEncoder];
+
+			for (uint32_t imgRgnIdx = 0; imgRgnIdx < pCopyImageToMemoryInfo->regionCount; imgRgnIdx++) {
+				auto& imgRgn = pCopyImageToMemoryInfo->pRegions[imgRgnIdx];
+				auto& imgSubRez = imgRgn.imageSubresource;
+				id<MTLTexture> mtlTex = getMTLTexture(getPlaneFromVkImageAspectFlags(imgSubRez.aspectMask));
+				for (uint32_t imgLyrIdx = 0; imgLyrIdx < imgSubRez.layerCount; imgLyrIdx++) {
+					[mtlBlitEnc synchronizeTexture: mtlTex
+											 slice: imgSubRez.baseArrayLayer + imgLyrIdx
+											 level: imgSubRez.mipLevel];
+				}
+			}
+
+			[mtlBlitEnc endEncoding];
+			[mtlCmdBuff commit];
+			[mtlCmdBuff waitUntilCompleted];
+		}
+	}
+#endif
```

Comment:
I wonder if there's a point where we should be calling `-synchronizeTexture:slice:level:` where we aren't in response to a barrier command.

Reply:
> I wonder if there's a point where we should be calling `-synchronizeTexture:slice:level:` where we aren't in response to a barrier command.

I'm not sure what you are wondering here?

Reply:
I mean, the host has to synchronize its use of the image with the device before calling this function. According to [the Vulkan spec][1]:

> `vkCopyImageToMemoryEXT` does not check whether the device memory associated with `srcImage` is currently in use before performing the copy. The application **must** guarantee that any previously submitted command that writes to the copy regions has completed before the host performs the copy.

It's unclear, though, if that means that the host is responsible for emitting a barrier, or if this function includes an implicit barrier. I wonder, in particular, if the `vkTransitionImageLayoutEXT()` function has something to do with this.

[1]: https://registry.khronos.org/vulkan/specs/1.3-extensions/html/chap20.html#vkCopyImageToMemoryEXT

Reply:
> It's unclear, though, if that means that the host is responsible for emitting a barrier

Yes. I was unsure about that spec wording too. 

The spec wording for `vkCopyImageToMemoryEXT()` also says

> This command (] is functionally similar to `vkCmdCopyImageToBuffer2, except it is executed on the host and writes to host memory instead of a buffer.

which, if you squint hard enough, might imply the same barrier operations apply. But the spec is silent on which barriers to use.

Some of the CTS tests failing on Radeon did indeed emit a pipeline image barrier after rendering, but they only included `VK_ACCESS_TRANSFER_READ_BIT`, and not the `VK_ACCESS_HOST_READ_BIT` that I would have expected when preparing for a host-copy.

I actually did initially experiment with modifying `MVKImageMemoryBinding::needsHostReadSync()`, to force a sync on `VK_ACCESS_TRANSFER_READ_BIT` if the image included `VK_IMAGE_USAGE_HOST_TRANSFER_BIT_EXT`. It did solve the issue, but I was concerned that it would unduly penalize any non-host transfers that the image might be involved with. I wasn't convinced that setting `VK_IMAGE_USAGE_HOST_TRANSFER_BIT_EXT` implies that the app will focus on host-copy transfers only.

My other consideration was that since `vkCopyImageToMemoryEXT()` is not a GPU command, there's an implicit understanding that the CPU is expecting to wait for the GPU to be finished anyway, and so taking the hit when we know it's expected, was more appropriate.

`vkTransitionImageLayoutEXT()` is about layouts, and not memory access, and I couldn't find anything equivalent in Metal. Plus the problematic CTS tests were not issuing a call to that function anyway.

So, we could follow the path of forcing a sync on every `VK_ACCESS_TRANSFER_READ_BIT` barrier, if that is more appropriate, but the spec is definitely not clear on what's expected.